proof/refine/ARM/IpcCancel_R.thy

```diff
@@ -2660,6 +2660,106 @@ proof -
   done
 qed
 
+lemma replyUnlink_valid_tcb_state'_restart[wp]:
+  "\<lbrace>\<lambda>s. sch_act_wf (ksSchedulerAction s) s \<and> sch_act_simple s\<rbrace>
+   replyUnlink r t
+   \<lbrace>\<lambda>_ s. sch_act_wf (ksSchedulerAction s) s\<rbrace>"
+  apply (clarsimp simp: replyUnlink_def setReplyTCB_def getReplyTCB_def liftM_def)
+  by (wpsimp wp: hoare_drop_imp)
+
+lemma replyUnlink_weak_sch_act_wf[wp]:
+  "\<lbrace>\<lambda>s. weak_sch_act_wf (ksSchedulerAction s) s \<and> sch_act_not t s\<rbrace>
+   replyUnlink r t
+   \<lbrace>\<lambda>_ s. weak_sch_act_wf (ksSchedulerAction s) s\<rbrace>"
+  unfolding replyUnlink_def setReplyTCB_def getReplyTCB_def
+  by (wpsimp wp: hoare_vcg_imp_lift hoare_vcg_all_lift gts_wp'
+              simp: weak_sch_act_wf_def)
+
+lemma replyUnlink_list_refs_of_replies'[wp]:
+  "replyUnlink r t \<lbrace>\<lambda>s. P (list_refs_of_replies' s)\<rbrace>"
+  unfolding replyUnlink_def setReplyTCB_def getReplyTCB_def
+  apply (wpsimp simp: updateObject_default_def setObject_def split_def wp: gts_wp')
+  apply (erule arg_cong[where f=P, THEN iffD1, rotated])
+  apply (rule ext)
+  apply (clarsimp simp: opt_map_def sym_refs_def fun_upd_def list_refs_of_reply'_def
+                        map_set_def projectKO_opt_reply obj_at'_real_def ko_wp_at'_def
+                 split: option.split)
+  done
+
+lemma replyUnlink_valid_pspace'[wp]:
```

Comment:
We can get the validity of a reply object from `valid_pspace’` here (you can get `valid_objs'` from `valid_pspace’`) . So, we only need `valid_pspace’`  as the precondition. The rest should be derivable. The same applies to `replyUnlink_valid_idle’` below, too.